aten/src/ATen/CPUApplyUtils.h

```diff
@@ -22,289 +22,333 @@ namespace at {
+          strides[TH_TENSOR_dim_index] = tensor.strides()[i];
+        }
+      } /* Size of the inner most section */
+      size = sizes[dim - 1]; /* Stride of the inner most section */
+      stride = strides[dim - 1];
+    }
+    i = 0;
+  }
+  ~util_tensor() {
+    if (counter != NULL)
+      delete[] counter;
+  }
+};
+
+template <typename T>
+static bool tensor_update_counters(util_tensor<T>& ut) {
+  if (ut.i == ut.size) {
+    if (ut.contiguous)
+      return true;
 
-// TODO: turn this macro into a proper template
-#define  __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(ATENSOR, ALWAYS_UPDATE) \
-  if(ATENSOR##_i == ATENSOR##_size || ALWAYS_UPDATE) \
-  { \
-    if(ATENSOR##_contiguous) \
-      break; \
-\
-    if(ATENSOR##_dim == 1) \
-       break; \
-\
-    /* Reset pointer to beginning of loop */ \
-    ATENSOR##_data -= ATENSOR##_size*ATENSOR##_stride; \
-    for(ATENSOR##_i = ATENSOR##_dim-2; ATENSOR##_i >= 0; ATENSOR##_i--) \
-    { \
-      ATENSOR##_counter[ATENSOR##_i]++; \
-      /* Jump ahread by the stride of this dimension */ \
-      ATENSOR##_data += ATENSOR##_strides[ATENSOR##_i]; \
-\
-      if(ATENSOR##_counter[ATENSOR##_i]  == ATENSOR##_sizes[ATENSOR##_i]) \
-      { \
-        if(ATENSOR##_i == 0) \
-        { \
-          TH_TENSOR_APPLY_hasFinished = true; \
-          break; \
-        } \
-          else \
-        { \
-          /* Reset the pointer to the beginning of the chunk defined by this dimension */ \
-          ATENSOR##_data -= ATENSOR##_counter[ATENSOR##_i]*ATENSOR##_strides[ATENSOR##_i]; \
-          ATENSOR##_counter[ATENSOR##_i] = 0; \
-        } \
-      } \
-      else \
-        break; \
-    } \
-    ATENSOR##_i = 0; \
+    if (ut.dim == 1)
+      return true;
+
+    /* Reset pointer to beginning of loop */
+    ut.data -= ut.size * ut.stride;
+    for (ut.i = ut.dim - 2; ut.i >= 0; ut.i--) {
+      ut.counter[ut.i]++;
+      /* Jump ahread by the stride of this dimension */
+      ut.data += ut.strides[ut.i];
+
+      if (ut.counter[ut.i] == ut.sizes[ut.i]) {
+        if (ut.i == 0) {
+          return true;
+        } else {
+          /* Reset the pointer to the beginning of the chunk defined by this
+           * dimension */
+          ut.data -= ut.counter[ut.i] * ut.strides[ut.i];
+          ut.counter[ut.i] = 0;
+        }
+      } else
+        return false;
+    }
+    ut.i = 0;
   }
+  return false;
+}
 
 template <typename scalar1, typename scalar2, typename Op>
-void CPU_tensor_apply2_dim(Tensor& tensor1, Tensor& tensor2, int64_t dim, Op op) {
+void CPU_tensor_apply2_dim(
+    Tensor& tensor1,
+    Tensor& tensor2,
+    int64_t dim,
+    Op op) {
   checkBackend("CPU_tensor_apply2", {tensor1, tensor2}, Backend::CPU);
-  bool TH_TENSOR_APPLY_hasFinished = false;
-  int64_t TH_TENSOR_dim_index = 0;
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar1, tensor1, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar2, tensor2, dim, 1)
+  util_tensor<scalar1> ut1(tensor1, dim, true);
```

Comment:
`util_tensor` doesn't really tell you anything what this thing even is. How about `strided_tensor_iter`? We're really using it to traverse the elements of the tensor in a certain order, so it's an iterator